chips/sam4l/src/chip.rs

```diff
@@ -152,4 +153,16 @@ impl Chip for Sam4l {
     fn systick(&self) -> &cortexm4::systick::SysTick {
         self.systick
     }
+
+    fn prepare_for_sleep(&self) {
+        if pm::deep_sleep_ready() {
+            unsafe {
+                cortexm4::scb::set_sleepdeep();
+            }
+        } else {
+            unsafe {
+                cortexm4::scb::unset_sleepdeep();
```

Comment:
Correct. The way it works is you always call WFI (or WFE) to go into any sleep states. The chip consults various registers to figure out which state to go into---the DEEP SLEEP bit in the ARM core determines if it goes into WAIT/RETENTION (choosing between those is determined by a different SAM4L registers).